src/Rings/MPolyMap/AffineAlgebras.jl

```diff
@@ -212,11 +213,22 @@ function has_preimage_with_preimage(F::AffAlgHom, f::Union{MPolyRingElem, MPolyQ
 end
 
 @doc raw"""
-    preimage(F::AffAlgHom, I::U) where U <: Union{MPolyIdeal, MPolyQuoIdeal}
+    preimage(F::MPolyAnyMap, I::Ideal)
 
 Return the preimage of the ideal `I` under `F`.
 """
-function preimage(f::AffAlgHom, I::Union{MPolyIdeal, MPolyQuoIdeal})
+function preimage(F::MPolyAnyMap, I::Ideal)
+  # This generic routine does not work for maps where the domain is a quotient ring. 
+  # Hence it has been split into two specialized methods below.
+  error("not implemented")
+end
+
+function preimage(
+    f::MPolyAnyMap{<:MPolyRing{T}, CT}, 
+    I::Union{MPolyIdeal, MPolyQuoIdeal}
+  ) where {T <: FieldElem,
+           U2 <: MPolyRingElem{T}, 
+           CT <: Union{MPolyRing{T}, MPolyQuoRing{U2}}}
```

Comment:
```suggestion
           CT <: Union{MPolyRing{T}, MPolyQuoRing{<:MPolyRingElem{T}}}}
```
seems to make Aqua happy.

Reply:
Yes, good catch! The problem with the initial version is that there are inputs where the type variable `U2` has no restrictions imposed. Indeed this is the case if `CT` is a `MPolyRing{T}` and thus matches the first part of the union.